File: AdvancedLevel_C/1057.py
# Python 怎么也过不了，算了！
maxn, top = 100010, 0
stack = [0 for i in range(maxn)]
sums = [0 for i in range(maxn)]
lowbit = lambda n: n & (- n)

def update(i, v):
    while i < maxn:
        sums[i] += v
        i += lowbit(i)

def getsum(i):
    sum = 0
    while i >= 1:
        sum += sums[i]
        i -= lowbit(i)
    return sum

def median():
    left, right, k = 1, maxn, (top + 1) >> 1
    while left < right:
        mid = (left + right) >> 1
        if(getsum(mid) >= k):
            right = mid
        else:
            left = mid + 1
    return left


for i in range(int(input())):
    command = input()
    if command[1] == 'e':
        if top > 0:
            print(median())
        else:
            print("Invalid")

    elif command[1] == 'o':
        if top > 0:
            top -= 1
            update(stack[top], -1)
            print(stack[top])
        else:
            print("Invalid")

    else:
        stack[top] = int(command[5:])
        update(stack[top], 1)
        top += 1

File: AdvancedLevel_C/1057.c
#include <stdio.h>
#include <string.h>

#define MAXN 100010
#define lowbit(i) ((i) & (-i))

int C[MAXN], stack[MAXN];

void update(int x, int v)
{
    for (int i = x; i < MAXN; i += lowbit(i))
        C[i] += v;
}

int getsum(int x)
{
    int sum = 0;
    for (int i = x; i >= 1; i -= lowbit(i))
        sum += C[i];
    return sum;
}
// O(logN ^ 2)
int search(int left, int right, int k)
{
    while (left < right) {
        int mid = (left + right) / 2;
        if (getsum(mid) >= k)
            right = mid;
        else
            left = mid + 1;
    }
    return left;
}

int main()
{
    int n, num, top = 0;
    char command[16];
#ifdef __TINYC__
    freopen("1057.txt", "w+", stdout);
#endif
    scanf("%d", &n);
    for (int i = 0; i < n; ++i) {
        scanf("%s", command);
        switch (command[1]) {
            case 'e':
            case 'o':
                if (top > 0) {
                    if (command[1] == 'o') {
                        num = stack[--top];
                        update(num, -1);
                    }
                    else
                        num = search(1, MAXN, (top + 1) / 2);
                    printf("%d\n", num);
                }
                else
                    printf("Invalid\n");
                break;
            case 'u':
                scanf("%d", &num);
                stack[top++] = num;
                update(num, 1);
                break;
        }
    }

    return 0;
}
